File: brique-102/packages/php/src/Http/HttpClient.php
<?php

namespace Molam\Http;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\RequestException;
use Molam\Exceptions\MolamException;

class HttpClient
{
    /** @var Client */
    private $client;

    /** @var int */
    private $maxRetries;

    /** @var array */
    private $options;

    public function __construct(array $options)
    {
        $this->options = $options;
        $this->maxRetries = $options['maxRetries'] ?? 3;

        $this->client = new Client([
            'base_uri' => $options['baseUrl'],
            'timeout' => ($options['timeoutMs'] ?? 8000) / 1000,
            'headers' => [
                'Authorization' => 'Bearer ' . $options['apiKey'],
                'Content-Type' => 'application/json',
                'User-Agent' => 'Molam-SDK-PHP/2.0'
            ]
        ]);
    }

    /**
     * Make GET request
     */
    public function get(string $path, array $options = []): array
    {
        return $this->requestWithRetry('GET', $path, $options);
    }

    /**
     * Make POST request
     */
    public function post(string $path, array $body = [], array $options = []): array
    {
        $options['json'] = $body;
        return $this->requestWithRetry('POST', $path, $options);
    }

    /**
     * Make PUT request
     */
    public function put(string $path, array $body = [], array $options = []): array
    {
        $options['json'] = $body;
        return $this->requestWithRetry('PUT', $path, $options);
    }

    /**
     * Make DELETE request
     */
    public function delete(string $path, array $options = []): array
    {
        return $this->requestWithRetry('DELETE', $path, $options);
    }

    /**
     * Request with automatic retries
     */
    private function requestWithRetry(string $method, string $path, array $options): array
    {
        // Add idempotency key
        if (!isset($options['headers']['Idempotency-Key'])) {
            $options['headers']['Idempotency-Key'] = $this->generateIdempotencyKey();
        }

        $attempt = 0;

        while (true) {
            try {
                $response = $this->client->request($method, $path, $options);
                $body = json_decode((string)$response->getBody(), true);
                return $body ?? [];
            } catch (RequestException $e) {
                $status = $e->getResponse() ? $e->getResponse()->getStatusCode() : null;

                if ($attempt >= $this->maxRetries || !$this->isRetryableStatus($status)) {
                    throw MolamException::fromGuzzleException($e);
                }

                // Backoff
                $wait = $this->backoff($attempt);
                usleep($wait * 1000);
                $attempt++;
            }
        }
    }

    private function isRetryableStatus(?int $status): bool
    {
        if ($status === null) {
            return true;
        }

        if ($status >= 500) {
            return true;
        }

        return in_array($status, [408, 429, 425], true);
    }

    private function backoff(int $attempt): int
    {
        $sequence = [200, 500, 1000, 2000, 5000];
        return $sequence[min($attempt, count($sequence) - 1)];
    }

    private function generateIdempotencyKey(): string
    {
        return bin2hex(random_bytes(16));
    }
}

File: brique-102/packages/python/molam/http_client.py
"""
HTTP client with retries and idempotency
"""

import time
import uuid
from typing import Optional, Dict, Any
import requests

from molam.exceptions import MolamError


class HttpClient:
    """HTTP client with retries and idempotency"""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout_ms: int = 8000,
        max_retries: int = 3
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout_ms / 1000.0
        self.max_retries = max_retries

        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
            "User-Agent": "Molam-SDK-Python/2.0"
        })

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make GET request"""
        return self._request_with_retry("GET", path, params=params)

    def post(
        self,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None
    ) -> Dict[str, Any]:
        """Make POST request"""
        return self._request_with_retry("POST", path, json_data=body, extra_headers=headers)

    def put(
        self,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None
    ) -> Dict[str, Any]:
        """Make PUT request"""
        return self._request_with_retry("PUT", path, json_data=body, extra_headers=headers)

    def delete(self, path: str) -> Dict[str, Any]:
        """Make DELETE request"""
        return self._request_with_retry("DELETE", path)

    def _request_with_retry(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        json_data: Optional[Dict[str, Any]] = None,
        extra_headers: Optional[Dict[str, str]] = None
    ) -> Dict[str, Any]:
        """Execute request with automatic retries"""
        headers = extra_headers or {}

        # Add idempotency key if not present
        if "Idempotency-Key" not in headers:
            headers["Idempotency-Key"] = str(uuid.uuid4())

        url = f"{self.base_url}{path}"
        attempt = 0

        while True:
            try:
                response = self.session.request(
                    method=method,
                    url=url,
                    params=params,
                    json=json_data,
                    headers=headers,
                    timeout=self.timeout
                )

                response.raise_for_status()
                return response.json() if response.text else {}

            except requests.exceptions.RequestException as e:
                status = e.response.status_code if e.response is not None else None

                if attempt >= self.max_retries or not self._is_retryable_status(status):
                    raise MolamError.from_requests_exception(e)

                # Backoff
                wait = self._backoff(attempt)
                time.sleep(wait / 1000.0)
                attempt += 1

    @staticmethod
    def _is_retryable_status(status: Optional[int]) -> bool:
        """Check if HTTP status is retryable"""
        if status is None:
            return True
        if status >= 500:
            return True
        return status in [408, 429, 425]

    @staticmethod
    def _backoff(attempt: int) -> int:
        """Calculate backoff delay in milliseconds"""
        sequence = [200, 500, 1000, 2000, 5000]
        return sequence[min(attempt, len(sequence) - 1)]
